File: 03- March/13- Symmetric Tree/13- Symmetric Tree (Noura Algohary).cpp
// Author: Noura Algohary
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isMirror(TreeNode* left_root, TreeNode* right_root)
    {
        if(!left_root && !right_root)
            return true;
        else if(!left_root || !right_root)
            return false;
        else
        {
            if(left_root->val == right_root->val)
                return true && isMirror(left_root->left, right_root->right) && isMirror(left_root->right, right_root->left);
        }
        return false;
    }
    bool isSymmetric(TreeNode* root) {
        return isMirror(root->left, root->right);
    }
};

File: 03- March/13- Symmetric Tree/13- Symmetric Tree (Noura Algohary).py
# Author: Noura Algohary
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isMirror(self, left_root, right_root):
        if(not left_root and not right_root):
            return True
        elif(not left_root or not right_root):
            return False
        else:
            if (left_root.val == right_root.val):
                return True and self.isMirror(left_root.left, right_root.right) and self.isMirror(left_root.right, right_root.left)

    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return self.isMirror(root.left, root.right)